include/mra/kernels/multiply.h

```diff
@@ -11,70 +11,105 @@ namespace mra {
-      // convert to function values, multiply function values, convert back to coefficients
-      // project both nodes one level up.
-
-      // convert coeffs to function values
-      transform(nodeA, phiT, r1, workspace);
-      transform(nodeB, phiT, r2, workspace);
-      const T scale = std::pow(T(2), T(0.5 * NDIM * key.level())) / std::sqrt(D.template get_volume<T>());
+      if (keyA.level() > keyB.level()){
+        fcube_for_mul(D, keyA, keyB.right_child(), nodeB, rcnodeB, phibar, phi, quad_x, K, workspace);
+        T scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyB.right_child().level())));
+        rcnodeB *= scale;
+        fcube_for_mul(D, keyA, keyB.left_child(), nodeB, lcnodeB, phibar, phi, quad_x, K, workspace);
+        scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyB.left_child().level())));
+        lcnodeB *= scale;
+        fcube_for_mul(D, keyA, keyA.right_child(), nodeA, rcnodeA, phibar, phi, quad_x, K, workspace);
+        scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyA.right_child().level())));
+        rcnodeA *= scale;
+        fcube_for_mul(D, keyA, keyA.left_child(), nodeA, lcnodeA, phibar, phi, quad_x, K, workspace);
+        scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyA.left_child().child.level())));
+        lcnodeA *= scale;
+      }
+      else if (keyB.level() <= keyB.level()){
+        fcube_for_mul(D, keyB, keyA.right_child(), nodeA, rcnodeA, phibar, phi, quad_x, K, workspace);
+        T scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyB.right_child().level())));
+        rcnodeA *= scale;
+        fcube_for_mul(D, keyB, keyA.left_child(), nodeA, lcnodeA, phibar, phi, quad_x, K, workspace);
+        scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyB.left_child().level())));
+        lcnodeA *= scale;
+        fcube_for_mul(D, keyB, keyB.right_child(), nodeB, rcnodeB, phibar, phi, quad_x, K, workspace);
+        scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyA.right_child().level())));
+        rcnodeB *= scale;
+        fcube_for_mul(D, keyB, keyB.left_child(), nodeB, lcnodeB, phibar, phi, quad_x, K, workspace);
+        scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyA.left_child().child.level())));
+        lcnodeB *= scale;
 
+        // fcube_for_mul() returns function values evaluated at quadrature points
+      }
       foreach_idx(nodeA, [&](size_type i) {
-          r[i] = scale * r1[i] * r2[i];
-      });
+        lcnodeA[i] = lcnodeA[i] * lcnodeB[i];
+        rcnodeA[i] = rcnodeA[i] * rcnodeB[i];
+    });
 
       // convert back to coeffs
-      transform(r, phibar, nodeR, workspace);
+      transform(lcnodeA, phibar, r1, workspace);
+      transform(rcnodeA, phibar, r2, workspace);
+
+      // compress the result to nodeR
```

Comment:
We compress here, but keep the scaling function coefficients instead of wavelet coefficients.